src/MemoryProtocol.js

```diff
@@ -0,0 +1,129 @@
+import invariant from 'invariant';
+
+import createPath from './utils/createPath';
+import ensureLocation from './utils/ensureLocation';
+
+const STATE_KEY = '@@farce/state';
+
+export default class MemoryProtocol {
+  constructor(initialLocation, { persistent = false } = {}) {
+    this._persistent = persistent;
+
+    const initialState = persistent ? this._loadState() : null;
+    if (initialState) {
+      this._stack = initialState.stack;
+      this._index = initialState.index;
+    } else {
+      this._stack = [ensureLocation(initialLocation)];
+      this._index = 0;
+    }
+
+    this._keyPrefix = Math.random()
+      .toString(36)
+      .slice(2, 8);
+    this._keyIndex = 0;
+
+    this._listener = null;
+  }
+
+  _loadState() {
+    try {
+      const { stack, index } = JSON.parse(
+        window.sessionStorage.getItem(STATE_KEY),
+      );
+
+      // Checking instanceof Array is okay because we make the array.
+      if (
+        stack instanceof Array &&
+        typeof index === 'number' &&
+        stack[index]
+      ) {
+        return { stack, index };
+      }
+    } catch (e) {} // eslint-disable-line no-empty
+
+    return null;
+  }
+
+  init(delta = 0) {
+    return {
+      ...this._stack[this._index],
+      action: 'POP',
+      index: this._index,
+      delta,
```

Comment:
So I originally did this sticking index and delta on location, but what's the value in doing so?

Reply:
in principle for consistency with browser protocol... e.g. if you want to animate differently based on whether delta is positive or negative

not sure if index has any purpose... really just from https://github.com/ReactTraining/history/issues/36 and https://github.com/ReactTraining/history/issues/334